Add perfmap debug directory entry to crossgen2 output as needed

hoyosjs opened the pull request:
This change in general contains:
- Only pass perfmap argument for Linux SPC. We were always passing it down (not all that important)
- Add PerfMap Debug Directory Entry spec to the COFF spec for SRM.
- Enhance CrossGen2 to emit PerfMap debug directory record. Tested that I can correlate entries as needed on Windows. 
For example the header in the perfmap is:
```
FFFFFFFF 00 6819a895-6568-e40a-b8dc-cc7f97c73524
FFFFFFFE 00 1
FFFFFFFD 00 Windows
FFFFFFFC 00 X64
FFFFFFFC 00 CoreRT
```
and the entries in the PE are:
```
--- PE FILE: E:\repos\runtime\artifacts\bin\coreclr\windows.x64.Debug\System.Private.CoreLib.dll ---
CodeView
        System.Private.CoreLib.ni.pdb, Age = 1, 19c03592-b136-04bf-12f9-f1a7a2d1efc3
21
        System.Private.CoreLib.r2rmap, Signature = 95 a8 19 68 68 65 0a e4 b8 dc cc 7f 97 c7 35 24, Version = 1
CodeView
        E:\repos\runtime\artifacts\obj\coreclr\System.Private.CoreLib\x64\Debug\System.Private.CoreLib.pdb, Age = 1, 89aa3509-94b5-4f46-8f53-a0fab72a0639
Reproducible
```
- Stop emitting an ni.pdb record when PDB is not requested (Linux scenario).

hoyosjs committed:
Only pass perfmap argument for Linux

hoyosjs committed:
Add PerfMap Debug Directory Entry spec

hoyosjs commented:
cc: @tommcdon @mangod9 @brianrob

hoyosjs commented:
This is related to #58468 on main.

trylek commented:
In your snippet of the PerfMap file above,

> FFFFFFFC 00 X64
> FFFFFFFC 00 CoreRT

is the duplication of the key (-4) intentional?

trylek commented on src/coreclr/tools/aot/ILCompiler.ReadyToRun/CodeGen/ReadyToRunObjectWriter.cs:
```diff
@@ -322,26 +338,36 @@ public void EmitPortableExecutable()
+
+                    if (perfMapDebugDirectoryEntryNode is not null)
+                    {
+                        Debug.Assert(_generatePdbFile && _outputInfoBuilder is not null && _outputInfoBuilder.EnumerateInputAssemblies().Any());
```
Why do you need <code>_generatePdbFile</code> set when emitting perfmap directory node?

trylek replied:
[In other words, should this rather be <code>_generatePerfMapFile</code>?]

trylek commented on src/coreclr/tools/aot/ILCompiler.ReadyToRun/CodeGen/ReadyToRunObjectWriter.cs:
```diff
@@ -374,7 +400,7 @@ public void EmitPortableExecutable()
                             path = Path.GetDirectoryName(_objectFilePath);
                         }
-                        _symbolFileBuilder.SavePerfMap(path, _perfMapFormatVersion, _objectFilePath, _nodeFactory.Target.OperatingSystem, _nodeFactory.Target.Architecture);
+                        _symbolFileBuilder.SavePerfMap(path, _perfMapFormatVersion, _objectFilePath, _nodeFactory.Target);
```
Nice simplification, thank you!

trylek commented on src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/DebugDirectoryNode.cs:
```diff
@@ -114,6 +129,21 @@ public override ObjectData GetData(NodeFactory factory, bool relocsOnly = false)
                 builder.EmitReloc(entry, RelocType.IMAGE_REL_FILE_ABSOLUTE);
             }
 
+            // Second emit a record for the perfmap
```
Nit - at the first glance I would be inclined to put this code in <code>PerfMapDebugDirectoryEntryNode</code> as it deals with internals of the node like its length that's also declared there. But I see that it's not done that way for <code>NativeDebugDirectoryEntryNode</code> so I don't insist.

hoyosjs replied:
I can do that, I'll do it as an add-on cleanup commit that I can ignore when porting this to the release branch.

trylek commented on src/coreclr/tools/r2rdump/R2RDump.cs:
```diff
@@ -434,7 +434,9 @@ public void Dump(ReadyToRunReader r2r)
                     }
-                    PerfMapWriter.Write(perfmapPath, _options.PerfmapFormatVersion, ProduceDebugInfoMethods(r2r), ProduceDebugInfoAssemblies(r2r), r2r.TargetOperatingSystem, r2r.TargetArchitecture);
+                    // TODO: can't seem to find any place that surfaces the ABI. This is for debugging purposes, so may not be as relevant to be correct.
+                    TargetDetails details = new TargetDetails(r2r.TargetArchitecture, r2r.TargetOperatingSystem, TargetAbi.CoreRT);
```
This might merit @MichalStrehovsky's attention. I _think_ that CoreRT is basically the NativeAOT ABI which is different from CoreCLR ABI, I recall I had to make fixes around that in the early days of Crossgen2 development.

hoyosjs replied:
Interesting - that is the default ABI getting emitted currently (as least the NodeFactory's Target says so).

trylek reviewed (commented):
This mostly looks great to me, I believe I spotted a small copy & paste bug and I hope for feedback from Michal regarding the ABI but I guess you're the one who knows where this information is used. Thank you!

hoyosjs commented on docs/design/specs/PE-COFF.md:
```diff
@@ -107,3 +107,15 @@ When validating that Windows PDB matches the debug directory record check that t
 
 > Note that when the debugger (or other tool) searches for the PDB only GUID and Age fields are used to match the PDB, but the timestamp of the CodeView debug directory entry does not need to match the timestamp stored in the PDB. Therefore, to verify byte-for-byte identity of the PDB, the timestamp field should also be checked.
 
+### R2R PerfMap Debug Directory Entry (type 21)
```
I'll reference as it's a PE addition.

hoyosjs commented on src/coreclr/tools/aot/ILCompiler.Diagnostics/PerfMapWriter.cs:
```diff
@@ -72,6 +62,32 @@ public static void Write(string perfMapFileName, int perfMapFormatVersion, IEnum
+            perfMapWriter.WriteLine(CurrentFormatVersion.ToString(), (uint)PseudoRVA.FormatVersion, 0);
+            perfMapWriter.WriteLine(details.OperatingSystem.ToString(), (uint)PseudoRVA.TargetOS, 0);
+            perfMapWriter.WriteLine(details.Architecture.ToString(), (uint)PseudoRVA.TargetArchitecture, 0);
+            perfMapWriter.WriteLine(details.Abi.ToString(), (uint)PseudoRVA.TargetArchitecture, 0);
```
hehe Yes. Thanks.

MichalStrehovsky commented on src/coreclr/tools/aot/ILCompiler.Diagnostics/PerfMapWriter.cs:
```diff
@@ -72,6 +62,32 @@ public static void Write(string perfMapFileName, int perfMapFormatVersion, IEnum
+            // In particular, the order of it. Append only.
+            perfMapWriter.WriteLine(outputGuid.ToString(), (uint)PseudoRVA.OutputGuid, 0);
+            perfMapWriter.WriteLine(CurrentFormatVersion.ToString(), (uint)PseudoRVA.FormatVersion, 0);
+            perfMapWriter.WriteLine(details.OperatingSystem.ToString(), (uint)PseudoRVA.TargetOS, 0);
```
Do we have any safeguards that if someone decided to rename the enum values (for example because they're tired of seeing OSX because Apple decided to rebrand the OS back to MacOS), diagnostics doesn't become broken?

However much I hate duplication, these enums are internal to the type system and if we need to use them in a file format, they should be mapped to stable identifiers first.

hoyosjs replied:
You are right - I was intending to emit the numeric values now. However, even that seems unstable (reordering them or deleting values would mess things up).

davidwrighton replied:
The way we deal with that, is we define enums as part of the ReadyToRunConstants.cs file ... and then those are the enums we use. In general we don't change any of those values, we just add. (Although, we have a version field so we could change them if we really wanted to.) Please don't write code that assumes that the general TypeSystem constants are constant though. That's not OK, as we reserve the right to change them.

davidwrighton commented on src/coreclr/tools/Common/TypeSystem/Common/TargetArchitecture.cs:
```diff
@@ -9,13 +9,15 @@ namespace Internal.TypeSystem
-        X64,
-        X86,
-        Wasm32,
+        Unknown = 0,
```
Enums that are part of the TypeSystem are not considered to be part of the File format, and should not be considered to be stable. Please make separate enums in a file which is explicitly for the file format you are working with.

davidwrighton commented on src/coreclr/tools/Common/TypeSystem/Common/TargetDetails.cs:
```diff
@@ -9,37 +9,41 @@ namespace Internal.TypeSystem
-        NetBSD,
-        SunOS,
-        WebAssembly,
+        Unknown     = 0,
```
Enums that are part of the TypeSystem are not considered to be part of the File format, and should not be considered to be stable. Please make separate enums in a file which is explicitly for the file format you are working with.

davidwrighton commented on src/coreclr/tools/Common/TypeSystem/Common/TargetDetails.cs:
```diff
@@ -9,37 +9,41 @@ namespace Internal.TypeSystem
+    /// Deletions and renames are OK, additions are only allowed as appends.
+    public enum TargetAbi : uint
     {
-        Unknown,
```
Enums that are part of the TypeSystem are not considered to be part of the File format, and should not be considered to be stable. Please make separate enums in a file which is explicitly for the file format you are working with.

davidwrighton commented on src/coreclr/tools/aot/ILCompiler.Diagnostics/PerfMapWriter.cs:
```diff
@@ -16,13 +16,15 @@ public class PerfMapWriter
+        // These pseudo RVAs are used by other tools, please only append to this.
         public enum PseudoRVA : uint
         {
-            OutputGuid         = 0xFFFFFFFF,
```
I suggest moving this to a file which *only* contains constants that are expected to be preserved for file format purposes. However, as these constants are tied to the PerfMap format, if you do not wish to do so, that's alright too.

davidwrighton reviewed (changes requested):
Please do not attempt to stabilize any constants in the typesystem proper. That is not the appopriate place for stable constants.

MichalStrehovsky commented on src/coreclr/tools/aot/ILCompiler.Diagnostics/PerfMapWriter.cs:
```diff
@@ -72,6 +55,96 @@ public static void Write(string perfMapFileName, int perfMapFormatVersion, IEnum
+                _ => throw new NotImplementedException(details.OperatingSystem.ToString())
+            };
+
+            PerfMapAbiToken abiToken = details.Abi switch
+            {
+                TargetAbi.Unknown => PerfMapAbiToken.Unknown,
+                TargetAbi.CoreRT => PerfMapAbiToken.CoreRT,
+                TargetAbi.CoreRTArmel => PerfMapAbiToken.CoreRTArmel,
+                TargetAbi.Jit => PerfMapAbiToken.Jit,
+                TargetAbi.CppCodegen => PerfMapAbiToken.CppCodegen,
+                _ => throw new NotImplementedException(details.Abi.ToString())
```
I would take it as an opportunity to get rid of nonsensical names and never possible values.

1:1 mapping will be super confusing once we decide to clean up unused things and rename things that don't have a good name in the type system. Perfmap format is a stable format so it will be stuck with these decisions forever. They cannot be easily cleaned up.

```suggestion
            PerfMapAbiToken abiToken = details.Abi switch
            {
                TargetAbi.Unknown => PerfMapAbiToken.Unknown,
                TargetAbi.CoreRT => PerfMapAbiToken.Default,
                TargetAbi.CoreRTArmel => PerfMapAbiToken.Armel,
                _ => throw new NotImplementedException(details.Abi.ToString())
```

MichalStrehovsky commented on src/coreclr/tools/aot/ILCompiler.Diagnostics/PerfMapWriter.cs:
```diff
@@ -72,6 +55,96 @@ public static void Write(string perfMapFileName, int perfMapFormatVersion, IEnum
+                TargetArchitecture.ARM64 => PerfMapArchitectureToken.ARM64,
+                TargetArchitecture.X64 => PerfMapArchitectureToken.X64,
+                TargetArchitecture.X86 => PerfMapArchitectureToken.X86,
+                TargetArchitecture.Wasm32 => PerfMapArchitectureToken.Wasm32,
```
I would leave Wasm out of this until we decided whether a R2R perfmap for wasm even makes sense, which is pretty far out.

hoyosjs committed:
Enhance CrossGen2 to emit PerfMap debug directory entry

This is used to correlate PE's with their corresponding PerfMaps. For example the header in the perfmap could be:
```
FFFFFFFF 00 026D4D21B3EE3D93843FF7A964235822
FFFFFFFE 00 1
FFFFFFFD 00 1
FFFFFFFC 00 3
FFFFFFFB 00 1
```
And the PE will have the corresponding entries in the PE as:
```
PerfMap (Type 21):
        System.Private.CoreLib.ni.r2rmap, Signature = 026d4d21b3ee3d93843ff7a964235822, Version = 1
```

hoyosjs committed:
Emit headers for debug directory entries from the nodes themselves.

hoyosjs committed:
Extend perfmap and r2r documentation
